test/performance/odp_l2fwd.c

```diff
@@ -309,6 +320,211 @@ static inline int copy_packets(odp_packet_t *pkt_tbl, int pkts)
+{
+	odp_packet_t *pkt_tbl;
+	int i, pkts;
+
+	pkts = odp_packet_vector_tbl(vpkt, &pkt_tbl);
+	for (i = 0; i < pkts; i++)
+		odp_packet_free(pkt_tbl[i]);
+}
+
+/*
+ * Packet IO worker thread using scheduled queues and vector mode.
+ *
+ * arg  thread arguments of type 'thread_args_t *'
+ */
+static int run_worker_sched_mode_vector(void *arg)
+{
+	int thr;
+	int i;
+	int pktio, num_pktio;
+	uint16_t max_burst;
+	odp_pktout_queue_t pktout[MAX_PKTIOS];
+	odp_queue_t tx_queue[MAX_PKTIOS];
+	thread_args_t *thr_args = arg;
+	stats_t *stats = &thr_args->stats;
+	int use_event_queue = gbl_args->appl.out_mode;
+	pktin_mode_t in_mode = gbl_args->appl.in_mode;
+
+	thr = odp_thread_id();
+	max_burst = gbl_args->appl.burst_rx;
+
+	if (gbl_args->appl.num_groups) {
+		odp_thrmask_t mask;
+
+		odp_thrmask_zero(&mask);
+		odp_thrmask_set(&mask, thr);
+
+		/* Join non-default groups */
+		for (i = 0; i < thr_args->num_groups; i++) {
+			if (odp_schedule_group_join(thr_args->group[i],
+						    &mask)) {
+				ODPH_ERR("Join failed\n");
+				return -1;
+			}
+		}
+	}
+
+	num_pktio = thr_args->num_pktio;
+
+	if (num_pktio > MAX_PKTIOS) {
+		ODPH_ERR("Too many pktios %i\n", num_pktio);
+		return -1;
+	}
+
+	for (pktio = 0; pktio < num_pktio; pktio++) {
+		tx_queue[pktio] = thr_args->pktio[pktio].tx_queue;
+		pktout[pktio]   = thr_args->pktio[pktio].pktout;
+	}
+
+	printf("[%02i] PKTIN_SCHED_%s_VECTOR, %s\n", thr,
+	       (in_mode == SCHED_PARALLEL) ? "PARALLEL" :
+	       ((in_mode == SCHED_ATOMIC) ? "ATOMIC" : "ORDERED"),
+	       (use_event_queue) ? "PKTOUT_QUEUE" : "PKTOUT_DIRECT");
+
+	odp_barrier_wait(&gbl_args->init_barrier);
+
+	/* Loop packets */
+	while (!odp_atomic_load_u32(&gbl_args->exit_threads)) {
+		odp_event_t  ev_tbl[MAX_PKT_BURST];
+		int vpkts;
+
+		vpkts = odp_schedule_multi_no_wait(NULL, ev_tbl, max_burst);
+
+		if (vpkts <= 0)
+			continue;
+
+		for (i = 0; i < vpkts; i++) {
+			odp_packet_t pkt, *pkt_tbl;
+			odp_packet_vector_t vpkt;
+			int src_idx, dst_idx;
+			unsigned tx_drops;
+			int pkts, sent;
+
+			vpkt = NULL;
+			pkts = 0;
+			if (odp_likely(odp_event_type(ev_tbl[i]) == ODP_EVENT_PACKET_VECTOR)) {
+				vpkt = odp_packet_vector_from_event(ev_tbl[i]);
+				pkts = odp_packet_vector_tbl(vpkt, &pkt_tbl);
+			} else  { /* ODP_EVENT_PACKET */
+				pkt  = odp_packet_from_event(ev_tbl[i]);
+				pkt_tbl = &pkt;
+				pkts = 1;
+			}
+
+			if (odp_unlikely(gbl_args->appl.extra_feat)) {
+				if (gbl_args->appl.packet_copy) {
+					int fails;
+
+					fails = copy_packets(pkt_tbl, pkts);
+					stats->s.copy_fails += fails;
+				}
+
+				if (gbl_args->appl.chksum)
+					chksum_insert(pkt_tbl, pkts);
+
+				if (gbl_args->appl.error_check) {
+					int rx_drops;
+
+					/* Drop packets with errors */
+					rx_drops = drop_err_pkts(pkt_tbl, pkts);
+
+					if (odp_unlikely(rx_drops)) {
+						stats->s.rx_drops += rx_drops;
+						if (pkts == rx_drops)
+							continue;
+
+						pkts -= rx_drops;
+						if (vpkt)
```

Comment:
No need for this check if only packet vectors are accepted.

Reply:
Agree.